src/Compilers/Core/Portable/EmbeddedText.cs

```diff
@@ -0,0 +1,345 @@
+            int length = (int)stream.Length;
+
+            if (length < CompressionThreshold)
+            {
+                var builder = Cci.PooledBlobBuilder.GetInstance();
+                builder.WriteInt32(0);
+
+                int bytesWritten = builder.TryWriteBytes(stream, length);
+                if (length != bytesWritten)
+                {
+                    throw new EndOfStreamException();
+                }
+
+                return builder.ToImmutableArrayAndFree();
+            }
+            else
+            {
+                Debug.Assert(length > 0);
+                var builder = BlobBuildingStream.GetInstance();
+                builder.WriteInt32(length);
+
+                using (var deflater = new DeflateStream(builder, CompressionLevel.Optimal))
+                {
+                    stream.CopyTo(deflater);
+                }
+
+                return builder.ToImmutableArrayAndFree();
+            }
+        }
+
+        internal static ImmutableArray<byte> CreateBlob(byte[] buffer, int length)
+        {
+            Debug.Assert(buffer != null);
+            Debug.Assert(length > 0);
+            Debug.Assert(length <= buffer.Length);
+
+            if (length < CompressionThreshold)
+            {
+                var builder = Cci.PooledBlobBuilder.GetInstance();
+                builder.WriteInt32(0);
+                builder.WriteBytes(buffer, 0, length);
+                return builder.ToImmutableArrayAndFree();
+            }
+            else
+            {
+                Debug.Assert(length > 0);
+                var builder = BlobBuildingStream.GetInstance();
+                builder.WriteInt32(length);
+
+                using (var deflater = new DeflateStream(builder, CompressionLevel.Optimal))
+                {
+                    deflater.Write(buffer, 0, length);
+                }
+
+                return builder.ToImmutableArrayAndFree();
+            }
+        }
+
+        private static ImmutableArray<byte> CreateBlob(SourceText sourceText)
+        {
+            Debug.Assert(sourceText != null);
+            Debug.Assert(sourceText.CanBeEmbedded);
+            Debug.Assert(sourceText.Encoding != null);
+            Debug.Assert(sourceText.PrecomputedEmbeddedTextBlob.IsDefault);
+
+            int maxByteCount = int.MaxValue;
+            try { maxByteCount = sourceText.Encoding.GetMaxByteCount(sourceText.Length);  }
+            catch (ArgumentOutOfRangeException) { }
+
+            var builder = BlobBuildingStream.GetInstance();
+            builder.WriteInt32(0);
+
+            var deflater = maxByteCount > CompressionThreshold ? new CountingDeflateStream(builder) : null;
+            using (var writer = new StreamWriter((Stream)deflater ?? builder, sourceText.Encoding))
+            {
+                sourceText.Write(writer);
+            }
+
+            if (deflater != null)
+            {
+                builder.PatchLeadingInt32(deflater.BytesWritten);
+            }
```

Comment:
There is a lot of conditional code here. It would be more readable to split compressed and uncompressed cases:

```
if (maxByteCount > CompressionThreshold)
{
   // deflating
}
else
{
   // simple copy
}
```